solutions/usaco-624.mdx

```diff
@@ -2,285 +2,378 @@
 id: usaco-624
 source: USACO Platinum 2016 February
 title: Load Balancing
-author: Benjamin Qi
+author: Benjamin Qi, Chuyang Wang
 ---
 
-## Explanation
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_balancing_platinum_feb16.html)
 
-The
-[official solution](http://www.usaco.org/current/data/sol_balancing_platinum_feb16.html)
-uses a segment tree, which is introduced in Gold. However, we can just binary
-search on the answer.
+## Solution 1: With Segment Tree
 
-## Implementation
+### Explanation
+
+The problem can be solved by sweeping a line from bottom to top. This line
+divides the farm into two regions, north and south. At the beginning, all cows
+belong to the northern region. For each of these lines, we need an efficient way
+to determine the optimal west/east partition to divide the farm into four pieces
+with the maximum number of cows appearing in one of the four regions, $M$, being
+minimized. To achieve this, we store the positions of cows in the northern and
+southern regions in two separate segment trees. These two segment trees hold the
+number of cows on a certain $x$ coordinate which belong to one of the two
+regions, respectively. Then, we do a binary search on both segment trees to find
+the best vertical line which minimizes $M$. In particular, we begin from the root
+and always add either the left or right subtree to the western or eastern
+region, depending on which of these two options leads to the smallest increase
+of $M$. At the end, we update our overall best solution and sweep the line on
+step to the top. We add all cows with current $y$ coordinate to the southern
```

Comment:
"sweep the line on the step to the top"
maybe "move the line to the next y-coordinate of interest"